crystalfontz/dbus/domain/base.py

```diff
@@ -0,0 +1,213 @@
+from typing import (
+    ClassVar,
+    List,
+    Optional,
+    Protocol,
+    Self,
+    Type,
+    Union,
+)
+
+
+class TypeProtocol(Protocol):
+    t: ClassVar[str]
+
+
+def t(*args: Union[str, Type[TypeProtocol]]) -> str:
+    type_ = ""
+
+    for arg in args:
+        if isinstance(arg, str):
+            type_ += arg
+        else:
+            type_ += arg.t
+
+    return type_
+
+
+def struct(*args: Union[str, Type[TypeProtocol]]) -> str:
+    return t("(", *args, ")")
+
+
+def array(of: Union[str, Type[TypeProtocol]]) -> str:
+    return f"a{t(of)}"
+
+
+class NoneM:
+    t: ClassVar[str] = ""
+
+
+OptUintT = int
+
+
+class OptUintM:
+    """
+    Map optional positive integers to and from dbus types.
+
+    This class treats values less than 0 (ie, -1) as representing None.
+    """
+
+    t: ClassVar[str] = "i"
+    none: ClassVar[OptUintT] = -1
+
+    @staticmethod
+    def unpack(r: OptUintT) -> Optional[int]:
+        return r if r >= 0 else None
+
+    @classmethod
+    def pack(cls: Type[Self], r: Optional[int]) -> OptUintT:
+        return r if r is not None else cls.none
+
+
+OptPosFloatT = float
+
+
+class OptPosFloatM:
```

Comment:
I'm not thrilled by this naming convention. Should I write out `OptPositiveFloat` or simplify to `OptFloat`?

Reply:
This is only used for timeout and keypad brightness. I think I can name this `OptFloat` internally, knowing that it only represents positive floats, and expose a `KeypadBrightnessM`/`KeypadBrightnessT` in the public API.

Reply:
I split the difference - `KeypadBrightnessM`/`KeypadBrightnessT` are implemented, but `OptFloat` is still exposed for documentation purposes.